File: apps/mobile/ios/Sources/DesignSystem/Typography.swift
//
//  Typography.swift
//  TchatApp
//
//  Created by Claude on 21/09/2024.
//

import SwiftUI

#if canImport(UIKit)
import UIKit
#endif

/// Typography design tokens matching TailwindCSS v4 text scales
public struct Typography {

    // MARK: - Heading Styles

    /// Large heading style (32pt, bold)
    /// Maps to TailwindCSS: text-3xl font-bold
    public let headingLarge = Font.system(size: 32, weight: .bold, design: .default)

    /// Medium heading style (24pt, semibold)
    /// Maps to TailwindCSS: text-2xl font-semibold
    public let headingMedium = Font.system(size: 24, weight: .semibold, design: .default)

    /// Small heading style (20pt, semibold)
    /// Maps to TailwindCSS: text-xl font-semibold
    public let headingSmall = Font.system(size: 20, weight: .semibold, design: .default)

    // MARK: - Body Styles

    /// Large body text (18pt, regular)
    /// Maps to TailwindCSS: text-lg
    public let bodyLarge = Font.system(size: 18, weight: .regular, design: .default)

    /// Medium body text (16pt, regular) - Default body text
    /// Maps to TailwindCSS: text-base
    public let bodyMedium = Font.system(size: 16, weight: .regular, design: .default)

    /// Small body text (14pt, regular)
    /// Maps to TailwindCSS: text-sm
    public let bodySmall = Font.system(size: 14, weight: .regular, design: .default)

    // MARK: - Label Styles

    /// Caption text (12pt, medium)
    /// Maps to TailwindCSS: text-xs font-medium
    public let caption = Font.system(size: 12, weight: .medium, design: .default)

    /// Label text (14pt, medium)
    /// Maps to TailwindCSS: text-sm font-medium
    public let label = Font.system(size: 14, weight: .medium, design: .default)

    /// Button text (16pt, semibold)
    /// Maps to TailwindCSS: text-base font-semibold
    public let button = Font.system(size: 16, weight: .semibold, design: .default)

    // MARK: - Utility Styles

    /// Overline text (10pt, bold, uppercase)
    /// Maps to TailwindCSS: text-xs font-bold uppercase tracking-wide
    public let overline = Font.system(size: 10, weight: .bold, design: .default)

    /// Monospace text for code (14pt, regular, monospaced)
    /// Maps to TailwindCSS: text-sm font-mono
    public let code = Font.system(size: 14, weight: .regular, design: .monospaced)

    // MARK: - Accessibility Support

    /// Get scaled font for accessibility
    public func scaledFont(_ font: Font) -> Font {
        #if canImport(UIKit)
        // This would integrate with UIContentSizeCategory for dynamic type
        return font
        #else
        return font
        #endif
    }

    // MARK: - Line Heights

    /// Line height multipliers matching TailwindCSS leading values
    public struct LineHeight {
        public static let tight: CGFloat = 1.25     // leading-tight
        public static let normal: CGFloat = 1.5     // leading-normal
        public static let relaxed: CGFloat = 1.625  // leading-relaxed
        public static let loose: CGFloat = 2.0      // leading-loose
    }

    // MARK: - Letter Spacing

    /// Letter spacing values matching TailwindCSS tracking values
    public struct LetterSpacing {
        public static let tighter: CGFloat = -0.05  // tracking-tighter
        public static let tight: CGFloat = -0.025   // tracking-tight
        public static let normal: CGFloat = 0       // tracking-normal
        public static let wide: CGFloat = 0.025     // tracking-wide
        public static let wider: CGFloat = 0.05     // tracking-wider
        public static let widest: CGFloat = 0.1     // tracking-widest
    }
}

// MARK: - Font Weight Extensions

extension Font.Weight {
    /// TailwindCSS font weight mapping
    public static let thin = Font.Weight.ultraLight        // font-thin
    public static let extralight = Font.Weight.ultraLight  // font-extralight
    public static let light = Font.Weight.light            // font-light
    public static let normal = Font.Weight.regular         // font-normal
    public static let medium = Font.Weight.medium          // font-medium
    public static let semibold = Font.Weight.semibold      // font-semibold
    public static let bold = Font.Weight.bold              // font-bold
    public static let extrabold = Font.Weight.heavy        // font-extrabold
    public static let black = Font.Weight.black            // font-black
}

// MARK: - Typography Utilities

extension Typography {

    /// Get font for semantic usage
    public func fontFor(usage: TypographyUsage) -> Font {
        switch usage {
        case .pageTitle: return headingLarge
        case .sectionTitle: return headingMedium
        case .cardTitle: return headingSmall
        case .bodyText: return bodyMedium
        case .secondaryText: return bodySmall
        case .captionText: return caption
        case .buttonLabel: return button
        case .navigationTitle: return headingMedium
        case .tabLabel: return label
        case .inputLabel: return label
        case .errorText: return bodySmall
        }
    }
}

/// Semantic typography usage enum
public enum TypographyUsage {
    case pageTitle
    case sectionTitle
    case cardTitle
    case bodyText
    case secondaryText
    case captionText
    case buttonLabel
    case navigationTitle
    case tabLabel
    case inputLabel
    case errorText
}

File: apps/mobile/android/app/src/main/java/com/tchat/designsystem/Typography.kt
package com.tchat.designsystem

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

/**
 * Typography design tokens matching TailwindCSS v4 text scales
 */
object Typography {

    // Heading Styles

    /** Large heading style (32sp, bold)
     *  Maps to TailwindCSS: text-3xl font-bold */
    val headingLarge = TextStyle(
        fontSize = 32.sp,
        fontWeight = FontWeight.Bold,
        lineHeight = 40.sp,
        letterSpacing = (-0.25).sp
    )

    /** Medium heading style (24sp, semibold)
     *  Maps to TailwindCSS: text-2xl font-semibold */
    val headingMedium = TextStyle(
        fontSize = 24.sp,
        fontWeight = FontWeight.SemiBold,
        lineHeight = 32.sp,
        letterSpacing = 0.sp
    )

    /** Small heading style (20sp, semibold)
     *  Maps to TailwindCSS: text-xl font-semibold */
    val headingSmall = TextStyle(
        fontSize = 20.sp,
        fontWeight = FontWeight.SemiBold,
        lineHeight = 28.sp,
        letterSpacing = 0.sp
    )

    // Body Styles

    /** Large body text (18sp, regular)
     *  Maps to TailwindCSS: text-lg */
    val bodyLarge = TextStyle(
        fontSize = 18.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 28.sp,
        letterSpacing = 0.sp
    )

    /** Medium body text (16sp, regular) - Default body text
     *  Maps to TailwindCSS: text-base */
    val bodyMedium = TextStyle(
        fontSize = 16.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 24.sp,
        letterSpacing = 0.sp
    )

    /** Small body text (14sp, regular)
     *  Maps to TailwindCSS: text-sm */
    val bodySmall = TextStyle(
        fontSize = 14.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 20.sp,
        letterSpacing = 0.sp
    )

    // Label Styles

    /** Caption text (12sp, medium)
     *  Maps to TailwindCSS: text-xs font-medium */
    val caption = TextStyle(
        fontSize = 12.sp,
        fontWeight = FontWeight.Medium,
        lineHeight = 16.sp,
        letterSpacing = 0.4.sp
    )

    /** Label text (14sp, medium)
     *  Maps to TailwindCSS: text-sm font-medium */
    val label = TextStyle(
        fontSize = 14.sp,
        fontWeight = FontWeight.Medium,
        lineHeight = 20.sp,
        letterSpacing = 0.1.sp
    )

    /** Button text (16sp, semibold)
     *  Maps to TailwindCSS: text-base font-semibold */
    val button = TextStyle(
        fontSize = 16.sp,
        fontWeight = FontWeight.SemiBold,
        lineHeight = 24.sp,
        letterSpacing = 0.sp
    )

    // Utility Styles

    /** Overline text (10sp, bold, uppercase)
     *  Maps to TailwindCSS: text-xs font-bold uppercase tracking-wide */
    val overline = TextStyle(
        fontSize = 10.sp,
        fontWeight = FontWeight.Bold,
        lineHeight = 16.sp,
        letterSpacing = 1.5.sp
    )

    /** Monospace text for code (14sp, regular, monospaced)
     *  Maps to TailwindCSS: text-sm font-mono */
    val code = TextStyle(
        fontSize = 14.sp,
        fontWeight = FontWeight.Normal,
        lineHeight = 20.sp,
        letterSpacing = 0.sp,
        fontFamily = FontFamily.Monospace
    )

    // Line Height Constants

    /** Line height multipliers matching TailwindCSS leading values */
    object LineHeight {
        const val TIGHT = 1.25f     // leading-tight
        const val NORMAL = 1.5f     // leading-normal
        const val RELAXED = 1.625f  // leading-relaxed
        const val LOOSE = 2.0f      // leading-loose
    }

    // Letter Spacing Constants

    /** Letter spacing values matching TailwindCSS tracking values */
    object LetterSpacing {
        val TIGHTER = (-0.8).sp  // tracking-tighter
        val TIGHT = (-0.4).sp    // tracking-tight
        val NORMAL = 0.sp        // tracking-normal
        val WIDE = 0.4.sp        // tracking-wide
        val WIDER = 0.8.sp       // tracking-wider
        val WIDEST = 1.6.sp      // tracking-widest
    }
}

// Font Weight Extensions

/** TailwindCSS font weight mapping */
object FontWeights {
    val thin = FontWeight.Thin           // font-thin
    val extraLight = FontWeight.ExtraLight  // font-extralight
    val light = FontWeight.Light         // font-light
    val normal = FontWeight.Normal       // font-normal
    val medium = FontWeight.Medium       // font-medium
    val semiBold = FontWeight.SemiBold   // font-semibold
    val bold = FontWeight.Bold           // font-bold
    val extraBold = FontWeight.ExtraBold // font-extrabold
    val black = FontWeight.Black         // font-black
}

// Typography Utilities

/**
 * Semantic typography usage enum
 */
enum class TypographyUsage {
    PAGE_TITLE,
    SECTION_TITLE,
    CARD_TITLE,
    BODY_TEXT,
    SECONDARY_TEXT,
    CAPTION_TEXT,
    BUTTON_LABEL,
    NAVIGATION_TITLE,
    TAB_LABEL,
    INPUT_LABEL,
    ERROR_TEXT
}

/**
 * Get font for semantic usage
 */
fun Typography.fontFor(usage: TypographyUsage): TextStyle {
    return when (usage) {
        TypographyUsage.PAGE_TITLE -> headingLarge
        TypographyUsage.SECTION_TITLE -> headingMedium
        TypographyUsage.CARD_TITLE -> headingSmall
        TypographyUsage.BODY_TEXT -> bodyMedium
        TypographyUsage.SECONDARY_TEXT -> bodySmall
        TypographyUsage.CAPTION_TEXT -> caption
        TypographyUsage.BUTTON_LABEL -> button
        TypographyUsage.NAVIGATION_TITLE -> headingMedium
        TypographyUsage.TAB_LABEL -> label
        TypographyUsage.INPUT_LABEL -> label
        TypographyUsage.ERROR_TEXT -> bodySmall
    }
}

/**
 * Apply accessibility scaling to text style
 */
fun TextStyle.scaledForAccessibility(scaleFactor: Float = 1.0f): TextStyle {
    return this.copy(fontSize = this.fontSize * scaleFactor)
}